lib/ukfs-devfs: Introduce ukfs filesystem for special files under /dev

andreittr opened the pull request:
### Description of changes

This changeset introduces `ukfs-devfs`, a tiny pseudo-filesystem for Unikraft providing 2 things:
- a global anonymous ramfs root where libraries can register device-like special files, independent of VFS state
- the "devfs" ukfs driver, which returns an instance of the above root on `vopen()`, used for mounting on e.g. `/dev/`

As a first provider of dev special files, `posix-tty` has been patched to provide the `null`, `zero`, `stdin`, `stdout`, and `stderr` files.
Additional prerequisites to these features and tangential changes included:
- ukfs pseudofs driver template
- fix styling in `posix-tty`'s `Config.uk`
- fix a refleak bug in `posix-tty`

~~Depends on the following PR and has it included as squashed commit that will be removed on rebase, please ignore during review:~~
~~- https://github.com/unikraft/unikraft/pull/1609~~

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

`CONFIG_LIBPOSIX_TTY_DEV_PSEUDO` & `CONFIG_LIBPOSIX_TTY_DEV_STDIO`.

andreittr committed:
lib/ukfs: Introduce pseudofs driver template

This change introduces the ukfs pseudofs driver template, a header with
common boilerplate code for implementing pseudofses: false filesystems
that return a rebind of a singular root whenever opened.

Checkpatch-Ignore: MACRO_ARG_REUSE
Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/ukfs-devfs: Introduce devfs filesystem

This change introduces the devfs pseudo-filesystem, intended to provide
special files under /dev. Libraries providing these special files create
them using the exported devfs root, while consumers can access it as the
"devfs" ukfs filesystem.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/posix-tty: Fix style of Config.uk

This change corrects style inconsistencies in Config.uk.
No functional changes.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

mogasergiu commented on lib/posix-tty/pseudo.c:
```diff
@@ -180,3 +187,31 @@ const struct uk_file *uk_zerofile_create(void)
+	r = uk_fs_createat(uk_fs_devfs_root, "zero", 4, 0666, O_EXCL,
+			   (union uk_fs_create_target){ .file = &zero_file });
+	if (unlikely(PTRISERR(r))) {
+		uk_pr_err("Failed to create /dev/null: %d\n", PTR2ERR(r));
```
s/null/zero

Also, won't previously created `/dev/null` leak?

andreittr replied:
> s/null/zero

Oops.

> Also, won't previously created /dev/null leak?

I was relying on them being cleaned up by devfs' term function that releases the root. ~~But I have to check if term funcs run on init error or not.~~

edit: I checked and we do indeed [run term functions if init func fails](https://github.com/unikraft/unikraft/blob/staging/lib/ukboot/boot.c#L435); so we can safely rely on devfs' own cleanup here.

mogasergiu commented on lib/posix-tty/tty.c:
```diff
@@ -89,10 +129,14 @@ static int init_posix_tty(struct uk_init_ctx *ictx __unused)
 
 	r = 0;
+#if CONFIG_LIBPOSIX_TTY_DEV_STDIO
+	r = init_posix_tty_dev_stdio(in, out, out);
```
Probably would make more sense to have #elif so that we don't have an unnecessary `r = 0` above? :sweat_smile:
(I know the compiler would optimize it out anyway but visually looks unpleasant)

andreittr replied:
Sure, I've no strong feelings on which looks better, but an elif is definitely more explicit.

mogasergiu commented on lib/posix-tty/tty.c:
```diff
@@ -68,23 +68,31 @@ static int init_posix_tty(struct uk_init_ctx *ictx __unused)
 
-	return 0;
+	r = 0;
+out:
```
What about closing/releasing the fds created by `uk_fdtab_open_named` up until a certain point?

andreittr replied:
:thinking: I was working under the assumption that any open fds will be cleaned up when the fdtab goes down on system shutdown, although there is an argument to be made about `init_posix_tty` not failing with partial side-effects. But then again, if a failure of an init function is always fatal we might not need to worry about it. ~~Then again again, I'm not sure failure of an init function will run the term functions or just panic :smile: ~~

~~So yeah, it's probably safest to clean up fds on failure too :+1: .~~

edit: I checked and we do indeed [run term functions if init func fails](https://github.com/unikraft/unikraft/blob/staging/lib/ukboot/boot.c#L435); so we can safely rely on fdtab's cleanup removing any partial side-effects without needing bespoke rollback code here.

mogasergiu commented on lib/posix-tty/pseudo.c:
```diff
@@ -180,3 +187,31 @@ const struct uk_file *uk_zerofile_create(void)
+	return 0;
+}
+
+uk_rootfs_initcall_prio(init_posix_tty_dev_pseudo, 0x0, UK_FS_PRIO_FSAVAIL);
```
Why not release the created files upon shutdown? Is it because it would not longer matter?
(Ditto for tty.c)

andreittr replied:
On system shutdown devfs will release its root, which will in turn recursively drop references to everything it holds, releasing them in turn. That way libs don't need to cleanup the files they put in pseudofses (and thus needing to keep track of them in parallel with the filesystem), they just get cleaned up automatically by the fs on shutdown.

mogasergiu commented on lib/posix-tty/tty.c:
```diff
@@ -26,6 +32,40 @@
+	r = uk_fs_createat(uk_fs_devfs_root, "stderr", 4, 0222, O_EXCL,
+			   (union uk_fs_create_target){ .file = err });
+	if (unlikely(PTRISERR(r))) {
+		uk_pr_err("Failed to create /dev/stderr: %d\n", PTR2ERR(r));
```
Same question as for zero/null. Won't stdin/stdout leak?

andreittr commented:
Pushed update addressing comments & adding `exportsyms.uk` file to explicitly express developer intentions.

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

andreittr commented:
Pushed rebase to current staging following merging of all prerequisites.

andreittr commented:
Pushed minor change adding comment.

andreittr committed:
lib/ukfile-pseudo: Move files out of posix-tty

This change moves the null, void, and zero pseudofile implementations
out of posix-tty and into a separate library.
In addition, the API has been changed to export the raw file structures
with static lifetimes directly instead of providing refcounting getters.
This is for several reasons:
- pseudofiles have guaranteed static lifetimes, thus "borrowing" a
  reference should be free; dealing with refcounts adds overhead
- the ukfile API demands that any function returning a file reference
  have it be refcounted; making an exception here would weaken the rule.
  Exporting the raw file reference makes the lifetime and scope explicit

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/ukfile-pseudo: Add devfs support

This change adds support to create the "null" and "zero" files under
devfs at boot.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/posix-tty: Add devfs support

This change adds support to create the "stdin", "stdout", and "stderr"
files under devfs at boot.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr committed:
lib/posix-vfs-fstab: Add option to mount /dev

This change adds a Kconfig option along with dedicated code to mount
devfs on /dev after running any fstabs.

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr commented:
Pushed major update with 2 additions:
- Moved the pseudofiles out of `posix-tty` and into their own library
- Added code to fstab for mounting devfs on `/dev` at boot

mogasergiu commented on lib/ukfile-pseudo/pseudo.c:
```diff
@@ -147,36 +143,11 @@ static const struct uk_file void_file = {
 };
-
-
-const struct uk_file *uk_nullfile_create(void)
```
I would also mention in the commit message why these have been removed in favor of static lifetime global vars.

mogasergiu reviewed (approved):
Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

andreittr commented:
Pushed rebase to current staging.

andreittr committed:
lib/posix-tty: Export raw serial file

This change removes the refcounting getter for the singleton serial
file, rather directly exporting the serial file struct itself.
This brings the serial file in line with other singleton files, as well
as fixes a reference leak in init_posix_tty().

Signed-off-by: Andrei Tatar <ttr@unikraft.io>

andreittr commented:
Pushed functional update & addressing comments:
- Clarified the reasoning for exporting the raw singleton files in `ukfile-pseudo` instead of refcounting getters
- Changed the serial singleton file in `posix-tty` to similarly export the raw file
- Fixed a minor refleak bug of the serial file during `init_posix_tty`
- Fixed a bug where `init_posix_tty` would silently fail to run if `ukfs-devfs` is not enabled

michpappas commented:
Approved-by: Michalis Pappas <michalis@unikraft.io>
